src/upgrade.py

```diff
@@ -271,6 +271,7 @@ def _recover_multi_unit_cluster(self) -> None:
                             "Instance not yet back in the cluster."
                             f" Retry {attempt.retry_state.attempt_number}/{RECOVER_ATTEMPTS}"
                         )
+                        self.charm._mysql.start_group_replication()
                         raise Exception
```

Comment:
Not in theory, but yes for now as we start GR as a best efforts additionally to the background start and I do not want to change any charm logic (removing exception generation might have side effects). The goal is to release candidate to stable. Up 2 @paulomach to refactor this after the vacation using the new knowledge from the Jira ticket.

Reply:
i think we should retry the whole flow (retry check to see if unit is in the cluster) for which we need to raise an exception

if group_replication is running and we run `START GROUP_REPLICATION`, we get an error which will result in a warning being logged to the console in `start_group_replication` method -> so I dont think there will be issues with the current implementation

```
 > start group_replication;
ERROR: 3093: The START GROUP_REPLICATION command failed since the group is already running.
```

Reply:
similarly, when group replication is already starting, we will only print a warning

```
> start group_replication;
ERROR: 3663: The START GROUP_REPLICATION command encountered a failure. Another instance of START/STOP GROUP_REPLICATION command is executing.
```

Reply:
So, I'm ok with the proposed solution given the bespoke nature of upgrade.
There are other cases though where the self rejoin fails due to exhausted attempts that I would rather treat on more generic fashion (e.g. https://github.com/canonical/mysql-k8s-operator/issues/415).
My only question is how to properly detect that auto rejoin attempts are exhausted, so we do not overlap mysqld and charm.